packages/fxa-settings/src/pages/ResetPasswordRedesign/CompleteResetPasswordOtp/container.tsx

```diff
@@ -0,0 +1,207 @@
+    token,
+    accountResetToken,
+    emailToHashWith,
+    kB,
+    recoveryKeyId,
+  } = location.state as CompleteResetPasswordLocationState;
+
+  const hasConfirmedRecoveryKey = !!(
+    accountResetToken &&
+    email &&
+    kB &&
+    recoveryKeyId
+  );
+
+  const isResetWithoutRecoveryKey = !!(code && token);
+
+  // TODO in FXA-7890 handle case where hasAccountRecoveryKey could not be determined
+  // and show banner to use account recovery key
+  // add metrics to track if users ever click on the link
+
+  const handleNavigationWithRecoveryKey = () => {
+    navigate(`/reset_password_with_recovery_key_verified${location.search}`, {
+      replace: true,
+    });
+  };
+
+  const handleNavigationWithoutRecoveryKey = () => {
+    navigate(`/reset_password_verified${searchParams}`, {
+      replace: true,
+    });
+  };
+
+  const resetPasswordWithRecoveryKey = async (
+    accountResetToken: string,
+    emailToUse: string,
+    kB: string,
+    newPassword: string,
+    recoveryKeyId: string
+  ) => {
+    const options = {
+      accountResetToken,
+      emailToHashWith: emailToUse,
+      password: newPassword,
+      recoveryKeyId,
+      kB,
+    };
+
+    // TODO move out of account model?
+    // Reset with recovery key directly uses authClient instead of GQL
+    const accountResetData = await account.resetPasswordWithRecoveryKey(
+      options
+    );
+
+    return accountResetData;
+  };
+
+  const resetPasswordWithoutRecoveryKey = async (
+    code: string,
+    emailToUse: string,
+    newPassword: string,
+    token: string
+  ) => {
+    // TODO move out of account model?
+    const accountResetData: AccountResetData =
+      await account.completeResetPassword(
+        keyStretchExperiment.queryParamModel.isV2(config),
+        token,
+        code,
+        emailToUse,
+        newPassword,
+        accountResetToken
+      );
+    return accountResetData;
+  };
+
+  const notifyBrowserOfSignin = async (accountResetData: AccountResetData) => {
+    if (
+      isSyncDesktopV3Integration(integration) ||
+      isSyncOAuthIntegration(integration)
+    ) {
+      firefox.fxaLoginSignedInUser({
+        authAt: accountResetData.authAt,
+        email,
+        keyFetchToken: accountResetData.keyFetchToken,
+        sessionToken: accountResetData.sessionToken,
+        uid: accountResetData.uid,
+        unwrapBKey: accountResetData.unwrapBKey,
+        verified: accountResetData.verified,
+      });
+    }
+  };
+
+  const submitNewPassword = async (newPassword: string) => {
+    try {
+      // The `emailToHashWith` option is returned by the auth-server to let the front-end
+      // know what to hash the new password with. This is important in the scenario where a user
+      // has changed their primary email address. In this case, they must still hash with the
+      // account's original email because this will maintain backwards compatibility with
+      // how account password hashing works previously.
```

Comment:
Probably worth noting this is only the case for the v1 key stretch, we're using a random salt for v2.